src/bjm.c

```diff
@@ -0,0 +1,277 @@
+        Joblist *joblist;
+        bjmJobFunc func = waitForJob(&privdata, &joblist);
+
+        func(privdata);                     // Execute the callback
+        atomicDecr(joblist->job_count, 1);   // Decrement count AFTER callback finishes
+        atomicDecr(queuedJobCount, 1);
+        atomicIncr(processedJobCount, 1);
+    }
+
+    return NULL;
+}
+
+
+// API
+void bjmInit(int numThreads) {
+    if (threadCount == numThreads) return;  // Silently skip to support testing
+    assert(threadCount == 0);   // But don't allow changing the number of threads
+
+    functionsCount = 0;
+    functionsCapacity = INITIAL_FUNCTION_CAPACITY;
+
+    functions = zmalloc(functionsCapacity * sizeof(bjmJobFunc));
+    jobsByFunc = zmalloc(functionsCapacity * sizeof(Joblist*));
+
+    atomicSet(queuedJobCount, 0);
+    atomicSet(processedJobCount, 0);
+
+    mutexFifoInit(&activeJoblists);
+
+    pthread_cond_init(&wakeup_cond, NULL);
+
+    threadCount = numThreads;
+    threads = zmalloc(threadCount * sizeof(pthread_t));
+
+    pthread_attr_t attr;
+    size_t stacksize;
+    pthread_attr_init(&attr);
+    pthread_attr_getstacksize(&attr, &stacksize);
+    if (stacksize < BJM_THREAD_STACK_SIZE) stacksize = BJM_THREAD_STACK_SIZE;
+    pthread_attr_setstacksize(&attr, stacksize);
+
+    for (int i = 0;  i < threadCount;  i++) {
+        void *arg = (void*)(intptr_t)i;
+        if (pthread_create(&threads[i], &attr, pthreadFunction, arg) != 0) {
+            serverLog(LL_WARNING, "Fatal: Can't initialize background jobs.");
+            exit(1);
+        }
+    }
+}
+
+
+// API
+bjmJobFuncHandle bjmRegisterJobFunc(bjmJobFunc func) {
+    return getFuncIdx(func) + 1;    // +1 to avoid 0 (uninitialized static) being a valid value
+}
+
+
+// API
+void bjmSubmitJob(bjmJobFuncHandle funcHandle, void *privdata) {
+    Joblist *joblist = jobsByFunc[funcHandle - 1];
+
+    mutexFifoLock(&joblist->jobs);
+    fifoPush(joblist->jobs.fifo, privdata);
+    atomicIncr(joblist->job_count, 1);
+    atomicIncr(queuedJobCount, 1);
+    if (fifoLength(joblist->jobs.fifo) == 1) {
+        // Reader threads take the activeJobists lock before the joblist lock.  But this can't
+        //  cause deadlock because this joblist isn't in the active joblist yet.
+        mutexFifoLock(&activeJoblists);
+        fifoPush(activeJoblists.fifo, joblist);
+        mutexFifoUnlock(&activeJoblists);
+    }
+    pthread_cond_signal(&wakeup_cond);
+    mutexFifoUnlock(&joblist->jobs);
+}
+
+
+// API
+void bjmKillThreads(void) {
+    for (int i = 0;  i < threadCount;  i++) {
+        if (threads[i] == pthread_self()) continue;
+        if (pthread_cancel(threads[i]) == 0) {
+            int err = pthread_join(threads[i], NULL);
+            if (err == 0) {
+                serverLog(LL_WARNING, "BJM thread #%d terminated", i);
+            } else {
+                serverLog(LL_WARNING, "BJM thread #%d can not be joined: %s", i, strerror(err));
+            }
+        }
+    }
+}
+
+
+// API
```

Comment:
I'm not a fan of this convention, it's not adding that much and it's confusing outside of amazon contexts.